File: src/main/python/net/coderextreme/data/arc.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "title"
meta2.content = "arc.x3d"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "creator"
meta3.content = "Lost, Doug Sanden I think"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "generator"
meta4.content = "x3d-tidy V3.0.2, https://www.npmjs.com/package/x3d-tidy"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "generator"
meta5.content = "manual"

head1.children.append(meta5)

X3D0.head = head1
Scene6 = x3d.Scene()
ProtoDeclare7 = x3d.ProtoDeclare()
ProtoDeclare7.name = "point"
ProtoInterface8 = x3d.ProtoInterface()
field9 = x3d.field()
field9.accessType = "inputOutput"
field9.type = "SFVec3f"
field9.name = "translation"

ProtoInterface8.field.append(field9)

ProtoDeclare7.ProtoInterface = ProtoInterface8
ProtoBody10 = x3d.ProtoBody()
Transform11 = x3d.Transform(DEF="node")
Shape12 = x3d.Shape()
Appearance13 = x3d.Appearance()
Material14 = x3d.Material()
Material14.diffuseColor = [1,0,0]

Appearance13.material = Material14

Shape12.appearance = Appearance13
Sphere15 = x3d.Sphere()
Sphere15.radius = 0.1

Shape12.geometry = Sphere15

Transform11.children.append(Shape12)
PositionInterpolator16 = x3d.PositionInterpolator(DEF="PI1")
PositionInterpolator16.key = [0,1]
PositionInterpolator16.keyValue = [(0, 0, 0),(0, 5, 0)]

Transform11.children.append(PositionInterpolator16)
Script17 = x3d.Script(DEF="MB1")
field18 = x3d.field()
field18.accessType = "inputOutput"
field18.type = "SFVec3f"
field18.name = "translation"
field18.value = [50,50,0]

Script17.field.append(field18)
field19 = x3d.field()
field19.accessType = "inputOutput"
field19.type = "SFVec3f"
field19.name = "old"

Script17.field.append(field19)
field20 = x3d.field()
field20.accessType = "inputOnly"
field20.type = "SFTime"
field20.name = "set_location"

Script17.field.append(field20)
field21 = x3d.field()
field21.accessType = "inputOutput"
field21.type = "MFVec3f"
field21.name = "keyValue"
field21.value = [(0, 0, 0),(0, 5, 0)]

Script17.field.append(field21)

Script17.sourceCode = '''ecmascript:\n"+
"		function set_location(value) {\n"+
"                    old = translation;\n"+
"		    translation = new SFVec3f(Math.random()*10-5, Math.random()*10-5, Math.random()*10-5);\n"+
"                    keyValue = new MFVec3f([old, translation]);\n"+
"		    // Browser.println(keyValue);\n"+
"		}'''

Transform11.children.append(Script17)
TimeSensor22 = x3d.TimeSensor(DEF="CL1")
TimeSensor22.cycleInterval = 3
TimeSensor22.loop = True

Transform11.children.append(TimeSensor22)
IS23 = x3d.IS()
connect24 = x3d.connect()
connect24.nodeField = "translation"
connect24.protoField = "translation"

IS23.connect.append(connect24)

Transform11.IS = IS23

ProtoBody10.children.append(Transform11)
ROUTE25 = x3d.ROUTE()
ROUTE25.fromNode = "CL1"
ROUTE25.fromField = "cycleTime"
ROUTE25.toNode = "MB1"
ROUTE25.toField = "set_location"

ProtoBody10.children.append(ROUTE25)
ROUTE26 = x3d.ROUTE()
ROUTE26.fromNode = "CL1"
ROUTE26.fromField = "fraction_changed"
ROUTE26.toNode = "PI1"
ROUTE26.toField = "set_fraction"

ProtoBody10.children.append(ROUTE26)
ROUTE27 = x3d.ROUTE()
ROUTE27.fromNode = "MB1"
ROUTE27.fromField = "keyValue_changed"
ROUTE27.toNode = "PI1"
ROUTE27.toField = "set_keyValue"

ProtoBody10.children.append(ROUTE27)
ROUTE28 = x3d.ROUTE()
ROUTE28.fromNode = "PI1"
ROUTE28.fromField = "value_changed"
ROUTE28.toNode = "node"
ROUTE28.toField = "set_translation"

ProtoBody10.children.append(ROUTE28)

ProtoDeclare7.ProtoBody = ProtoBody10

Scene6.children.append(ProtoDeclare7)
ProtoDeclare29 = x3d.ProtoDeclare()
ProtoDeclare29.name = "x3dconnector"
ProtoInterface30 = x3d.ProtoInterface()
field31 = x3d.field()
field31.accessType = "initializeOnly"
field31.type = "SFNode"
field31.name = "startnode"

ProtoInterface30.field.append(field31)
field32 = x3d.field()
field32.accessType = "initializeOnly"
field32.type = "SFNode"
field32.name = "endnode"

ProtoInterface30.field.append(field32)
field33 = x3d.field()
field33.accessType = "inputOnly"
field33.type = "SFVec3f"
field33.name = "set_startpoint"

ProtoInterface30.field.append(field33)
field34 = x3d.field()
field34.accessType = "inputOnly"
field34.type = "SFVec3f"
field34.name = "set_endpoint"

ProtoInterface30.field.append(field34)

ProtoDeclare29.ProtoInterface = ProtoInterface30
ProtoBody35 = x3d.ProtoBody()
Group36 = x3d.Group()
Transform37 = x3d.Transform(DEF="trans")
Transform38 = x3d.Transform(DEF="rotscale")
Shape39 = x3d.Shape()
Appearance40 = x3d.Appearance()
Material41 = x3d.Material()
Material41.diffuseColor = [0.2,0.7,0.7]
Material41.transparency = 0.5

Appearance40.material = Material41

Shape39.appearance = Appearance40
Cylinder42 = x3d.Cylinder()
Cylinder42.radius = 0.05

Shape39.geometry = Cylinder42

Transform38.children.append(Shape39)

Transform37.children.append(Transform38)

Group36.children.append(Transform37)
Script43 = x3d.Script(DEF="S1")
field44 = x3d.field()
field44.accessType = "initializeOnly"
field44.type = "SFNode"
field44.name = "startnode"

Script43.field.append(field44)
field45 = x3d.field()
field45.accessType = "initializeOnly"
field45.type = "SFNode"
field45.name = "endnode"

Script43.field.append(field45)
field46 = x3d.field()
field46.accessType = "inputOutput"
field46.type = "SFNode"
field46.name = "position"
Transform47 = x3d.Transform(USE="trans")

field46.children.append(Transform47)

Script43.field.append(field46)
field48 = x3d.field()
field48.accessType = "inputOutput"
field48.type = "SFNode"
field48.name = "rotscale"
Transform49 = x3d.Transform(USE="rotscale")

field48.children.append(Transform49)

Script43.field.append(field48)
field50 = x3d.field()
field50.accessType = "inputOnly"
field50.type = "SFVec3f"
field50.name = "set_startpoint"

Script43.field.append(field50)
field51 = x3d.field()
field51.accessType = "inputOnly"
field51.type = "SFVec3f"
field51.name = "set_endpoint"

Script43.field.append(field51)

Script43.sourceCode = '''ecmascript:\n"+
"        function recompute(startpoint,endpoint){\n"+
"	    if (typeof endpoint === 'undefined') {\n"+
"		return;\n"+
"	    }\n"+
"            var dif = endpoint.subtract(startpoint);\n"+
"            var dist = dif.length()*0.5;\n"+
"            var dif2 = dif.multiply(0.5);\n"+
"            var norm = dif.normalize();\n"+
"            var transl = startpoint.add(dif2);\n"+
"	    if (typeof Quaternion !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new Quaternion.rotateFromTo(new SFVec3f(0.0,1.0,0.0), norm)\n"+
"		    };\n"+
"	    } else if (typeof SFRotation !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new SFRotation(new SFVec3f(0.0,1.0,0.0),norm)\n"+
"		    };\n"+
"	    } else {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl\n"+
"		    };\n"+
"	    }\n"+
"	}\n"+
"	function recompute_and_route(startpoint, endpoint) {\n"+
"	      var trafo = recompute(startpoint, endpoint);\n"+
"	      if (trafo) {\n"+
"		      position.translation = trafo.translation;\n"+
"		      rotscale.rotation = trafo.rotation;\n"+
"		      rotscale.scale = trafo.scale;\n"+
"	      }\n"+
"	}\n"+
"        function initialize(){\n"+
"            recompute_and_route(startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_startpoint(val,t){\n"+
"            recompute_and_route(val,endnode.translation);\n"+
"        }\n"+
"        function set_endpoint(val,t){\n"+
"            recompute_and_route(startnode.translation,val);\n"+
"        }'''
IS52 = x3d.IS()
connect53 = x3d.connect()
connect53.nodeField = "startnode"
connect53.protoField = "startnode"

IS52.connect.append(connect53)
connect54 = x3d.connect()
connect54.nodeField = "endnode"
connect54.protoField = "endnode"

IS52.connect.append(connect54)
connect55 = x3d.connect()
connect55.nodeField = "set_startpoint"
connect55.protoField = "set_startpoint"

IS52.connect.append(connect55)
connect56 = x3d.connect()
connect56.nodeField = "set_endpoint"
connect56.protoField = "set_endpoint"

IS52.connect.append(connect56)

Script43.IS = IS52

Group36.children.append(Script43)

ProtoBody35.children.append(Group36)

ProtoDeclare29.ProtoBody = ProtoBody35

Scene6.children.append(ProtoDeclare29)
Viewpoint57 = x3d.Viewpoint()
Viewpoint57.description = "Only Viewpoint"
Viewpoint57.position = [0,0,5]

Scene6.children.append(Viewpoint57)
Background58 = x3d.Background()
Background58.skyColor = [(0.4, 0.4, 0.4)]

Scene6.children.append(Background58)
ProtoInstance59 = x3d.ProtoInstance(DEF="G1")
ProtoInstance59.name = "point"
fieldValue60 = x3d.fieldValue()
fieldValue60.name = "translation"
fieldValue60.value = "0 0.9066665 0"

ProtoInstance59.fieldValue.append(fieldValue60)

Scene6.children.append(ProtoInstance59)
ProtoInstance61 = x3d.ProtoInstance(DEF="G2")
ProtoInstance61.name = "point"
fieldValue62 = x3d.fieldValue()
fieldValue62.name = "translation"
fieldValue62.value = "0 0.9066665 0"

ProtoInstance61.fieldValue.append(fieldValue62)

Scene6.children.append(ProtoInstance61)
ProtoInstance63 = x3d.ProtoInstance(DEF="G3")
ProtoInstance63.name = "point"
fieldValue64 = x3d.fieldValue()
fieldValue64.name = "translation"
fieldValue64.value = "0 0.9066665 0"

ProtoInstance63.fieldValue.append(fieldValue64)

Scene6.children.append(ProtoInstance63)
ProtoInstance65 = x3d.ProtoInstance(DEF="G4")
ProtoInstance65.name = "point"
fieldValue66 = x3d.fieldValue()
fieldValue66.name = "translation"
fieldValue66.value = "0 0.9066665 0"

ProtoInstance65.fieldValue.append(fieldValue66)

Scene6.children.append(ProtoInstance65)
ProtoInstance67 = x3d.ProtoInstance(DEF="connector1")
ProtoInstance67.name = "x3dconnector"
fieldValue68 = x3d.fieldValue()
fieldValue68.name = "startnode"
ProtoInstance69 = x3d.ProtoInstance(USE="G1")
ProtoInstance69.name = "point"

fieldValue68.children.append(ProtoInstance69)

ProtoInstance67.fieldValue.append(fieldValue68)
fieldValue70 = x3d.fieldValue()
fieldValue70.name = "endnode"
ProtoInstance71 = x3d.ProtoInstance(USE="G2")
ProtoInstance71.name = "point"

fieldValue70.children.append(ProtoInstance71)

ProtoInstance67.fieldValue.append(fieldValue70)

Scene6.children.append(ProtoInstance67)
ProtoInstance72 = x3d.ProtoInstance(DEF="connector2")
ProtoInstance72.name = "x3dconnector"
fieldValue73 = x3d.fieldValue()
fieldValue73.name = "startnode"
ProtoInstance74 = x3d.ProtoInstance(USE="G1")
ProtoInstance74.name = "point"

fieldValue73.children.append(ProtoInstance74)

ProtoInstance72.fieldValue.append(fieldValue73)
fieldValue75 = x3d.fieldValue()
fieldValue75.name = "endnode"
ProtoInstance76 = x3d.ProtoInstance(USE="G3")
ProtoInstance76.name = "point"

fieldValue75.children.append(ProtoInstance76)

ProtoInstance72.fieldValue.append(fieldValue75)

Scene6.children.append(ProtoInstance72)
ProtoInstance77 = x3d.ProtoInstance(DEF="connector3")
ProtoInstance77.name = "x3dconnector"
fieldValue78 = x3d.fieldValue()
fieldValue78.name = "startnode"
ProtoInstance79 = x3d.ProtoInstance(USE="G1")
ProtoInstance79.name = "point"

fieldValue78.children.append(ProtoInstance79)

ProtoInstance77.fieldValue.append(fieldValue78)
fieldValue80 = x3d.fieldValue()
fieldValue80.name = "endnode"
ProtoInstance81 = x3d.ProtoInstance(USE="G4")
ProtoInstance81.name = "point"

fieldValue80.children.append(ProtoInstance81)

ProtoInstance77.fieldValue.append(fieldValue80)

Scene6.children.append(ProtoInstance77)
ROUTE82 = x3d.ROUTE()
ROUTE82.fromNode = "G1"
ROUTE82.fromField = "translation_changed"
ROUTE82.toNode = "connector1"
ROUTE82.toField = "set_startpoint"

Scene6.children.append(ROUTE82)
ROUTE83 = x3d.ROUTE()
ROUTE83.fromNode = "G2"
ROUTE83.fromField = "translation_changed"
ROUTE83.toNode = "connector1"
ROUTE83.toField = "set_endpoint"

Scene6.children.append(ROUTE83)
ROUTE84 = x3d.ROUTE()
ROUTE84.fromNode = "G1"
ROUTE84.fromField = "translation_changed"
ROUTE84.toNode = "connector2"
ROUTE84.toField = "set_startpoint"

Scene6.children.append(ROUTE84)
ROUTE85 = x3d.ROUTE()
ROUTE85.fromNode = "G3"
ROUTE85.fromField = "translation_changed"
ROUTE85.toNode = "connector2"
ROUTE85.toField = "set_endpoint"

Scene6.children.append(ROUTE85)
ROUTE86 = x3d.ROUTE()
ROUTE86.fromNode = "G1"
ROUTE86.fromField = "translation_changed"
ROUTE86.toNode = "connector3"
ROUTE86.toField = "set_startpoint"

Scene6.children.append(ROUTE86)
ROUTE87 = x3d.ROUTE()
ROUTE87.fromNode = "G4"
ROUTE87.fromField = "translation_changed"
ROUTE87.toNode = "connector3"
ROUTE87.toField = "set_endpoint"

Scene6.children.append(ROUTE87)

X3D0.Scene = Scene6
f = open("../data/arc.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/arc.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/arc.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/cplusplus/net/coderextreme/data/arc.cpp
/* delete next 5 lines for Visual Studio */
#define FALSE false
#define TRUE true
#define WINAPI
#define AFX_EXT_CLASS
#define EXPORT32
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/CylinderExample/pch.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/CylinderExample/framework.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/XML_PARSER.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/X3DLib.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/Abstracts.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/Concretes.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/glMath.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter02/CylinderExample/include/glut.h"
int main(int argc, char ** argv) {
X3D* X3D0 = new X3D();
X3D0->setProfile("Immersive");
X3D0->setVersion("3.3");
head* head1 = new head();
meta* meta2 = new meta();
meta2->setName("title");
meta2->setContent("arc.x3d");
head1->addMeta(meta2);

meta* meta3 = new meta();
meta3->setName("creator");
meta3->setContent("Lost, Doug Sanden I think");
head1->addMeta(meta3);

meta* meta4 = new meta();
meta4->setName("generator");
meta4->setContent("manual");
head1->addMeta(meta4);

meta* meta5 = new meta();
meta5->setName("identifier");
meta5->setContent("https://coderextreme.net/X3DJSONLD/src/main/data/arc.x3d");
head1->addMeta(meta5);

meta* meta6 = new meta();
meta6->setName("description");
meta6->setContent("a generic proto to connect two objects");
head1->addMeta(meta6);

X3D0->setHead(head1);

Scene* Scene7 = new Scene();
Viewpoint* Viewpoint8 = new Viewpoint();
Viewpoint8->setPosition(new float[3]{0,0,5});
Viewpoint8->setDescription("Only Viewpoint");
Scene7->addChild(Viewpoint8);

Background* Background9 = new Background();
Background9->setSkyColor(new float[3]{0.4,0.4,0.4});
Scene7->addChild(Background9);

ProtoDeclare* ProtoDeclare10 = new ProtoDeclare();
ProtoDeclare10->setName("point");
ProtoInterface* ProtoInterface11 = new ProtoInterface();
field* field12 = new field();
field12->setName("translation");
field12->setAccessType("inputOutput");
field12->setType("SFVec3f");
field12->setValue("0 0 0");
ProtoInterface11->addField(field12);

ProtoDeclare10->setProtoInterface(ProtoInterface11);

ProtoBody* ProtoBody13 = new ProtoBody();
Transform* Transform14 = new Transform();
Transform14->setDEF("node");
IS* IS15 = new IS();
Connect* connect16 = new Connect();
connect16->setNodeField("translation");
connect16->setProtoField("translation");
IS15->addConnect(connect16);

Transform14->setIS(IS15);

Shape* Shape17 = new Shape();
Sphere* Sphere18 = new Sphere();
Sphere18->setRadius(0.1);
Shape17->setGeometry(Sphere18);

Appearance* Appearance19 = new Appearance();
Material* Material20 = new Material();
Material20->setDiffuseColor(new float[3]{1,0,0});
Appearance19->setMaterial(Material20);

Shape17->setAppearance(Appearance19);

Transform14->addChild(Shape17);

PositionInterpolator* PositionInterpolator21 = new PositionInterpolator();
PositionInterpolator21->setDEF("PI1");
PositionInterpolator21->setKey(new float[2]{0,1}, 2);
PositionInterpolator21->setKeyValue(new float[6]{0,0,0,0,5,0});
Transform14->addChild(PositionInterpolator21);

Script* Script22 = new Script();
Script22->setDEF("MB1");
field* field23 = new field();
field23->setName("translation");
field23->setAccessType("inputOutput");
field23->setType("SFVec3f");
field23->setValue("50 50 0");
Script22->addField(field23);

field* field24 = new field();
field24->setName("old");
field24->setAccessType("inputOutput");
field24->setType("SFVec3f");
field24->setValue("0 0 0");
Script22->addField(field24);

field* field25 = new field();
field25->setName("set_location");
field25->setAccessType("inputOnly");
field25->setType("SFTime");
Script22->addField(field25);

field* field26 = new field();
field26->setName("keyValue");
field26->setAccessType("inputOutput");
field26->setType("MFVec3f");
field26->setValue("0 0 0 0 5 0");
Script22->addField(field26);


Script22.setSourceCode(`ecmascript:\n"+
"		function set_location(value) {\n"+
"                    old = translation;\n"+
"		    translation = new SFVec3f(Math.random()*10-5, Math.random()*10-5, Math.random()*10-5);\n"+
"                    keyValue = new MFVec3f([old, translation]);\n"+
"		    // Browser.println(keyValue);\n"+
"		}`)
Transform14->addChild(Script22);

TimeSensor* TimeSensor27 = new TimeSensor();
TimeSensor27->setDEF("CL1");
TimeSensor27->setCycleInterval(3);
TimeSensor27->setLoop(True);
Transform14->addChild(TimeSensor27);

ROUTE* ROUTE28 = new ROUTE();
ROUTE28->setFromNode("CL1");
ROUTE28->setFromField("cycleTime");
ROUTE28->setToNode("MB1");
ROUTE28->setToField("set_location");
Transform14->addChild(ROUTE28);

ROUTE* ROUTE29 = new ROUTE();
ROUTE29->setFromNode("CL1");
ROUTE29->setFromField("fraction_changed");
ROUTE29->setToNode("PI1");
ROUTE29->setToField("set_fraction");
Transform14->addChild(ROUTE29);

ROUTE* ROUTE30 = new ROUTE();
ROUTE30->setFromNode("MB1");
ROUTE30->setFromField("keyValue");
ROUTE30->setToNode("PI1");
ROUTE30->setToField("keyValue");
Transform14->addChild(ROUTE30);

ROUTE* ROUTE31 = new ROUTE();
ROUTE31->setFromNode("PI1");
ROUTE31->setFromField("value_changed");
ROUTE31->setToNode("node");
ROUTE31->setToField("set_translation");
Transform14->addChild(ROUTE31);

ProtoBody13->addChild(Transform14);

ProtoDeclare10->setProtoBody(ProtoBody13);

Scene7->addChild(ProtoDeclare10);

ProtoDeclare* ProtoDeclare32 = new ProtoDeclare();
ProtoDeclare32->setName("x3dconnector");
ProtoInterface* ProtoInterface33 = new ProtoInterface();
field* field34 = new field();
field34->setName("startnode");
field34->setAccessType("initializeOnly");
field34->setType("SFNode");
ProtoInterface33->addField(field34);

field* field35 = new field();
field35->setName("endnode");
field35->setAccessType("initializeOnly");
field35->setType("SFNode");
ProtoInterface33->addField(field35);

field* field36 = new field();
field36->setName("set_startpoint");
field36->setAccessType("inputOnly");
field36->setType("SFVec3f");
ProtoInterface33->addField(field36);

field* field37 = new field();
field37->setName("set_endpoint");
field37->setAccessType("inputOnly");
field37->setType("SFVec3f");
ProtoInterface33->addField(field37);

ProtoDeclare32->setProtoInterface(ProtoInterface33);

ProtoBody* ProtoBody38 = new ProtoBody();
Group* Group39 = new Group();
Transform* Transform40 = new Transform();
Transform40->setDEF("trans");
Transform* Transform41 = new Transform();
Transform41->setDEF("rotscale");
Shape* Shape42 = new Shape();
Appearance* Appearance43 = new Appearance();
Material* Material44 = new Material();
Material44->setDiffuseColor(new float[3]{0.2,0.7,0.7});
Material44->setTransparency(0.5);
Appearance43->setMaterial(Material44);

Shape42->setAppearance(Appearance43);

Cylinder* Cylinder45 = new Cylinder();
Cylinder45->setRadius(0.05);
Shape42->setGeometry(Cylinder45);

Transform41->addChild(Shape42);

Transform40->addChild(Transform41);

Group39->addChild(Transform40);

Script* Script46 = new Script();
Script46->setDEF("S1");
field* field47 = new field();
field47->setName("startnode");
field47->setAccessType("initializeOnly");
field47->setType("SFNode");
Script46->addField(field47);

field* field48 = new field();
field48->setName("endnode");
field48->setAccessType("initializeOnly");
field48->setType("SFNode");
Script46->addField(field48);

field* field49 = new field();
field49->setName("position");
field49->setAccessType("inputOutput");
field49->setType("SFNode");
Transform* Transform50 = new Transform();
Transform50->setUSE("trans");
field49->addChild(Transform50);

Script46->addField(field49);

field* field51 = new field();
field51->setName("rotscale");
field51->setAccessType("inputOutput");
field51->setType("SFNode");
Transform* Transform52 = new Transform();
Transform52->setUSE("rotscale");
field51->addChild(Transform52);

Script46->addField(field51);

field* field53 = new field();
field53->setName("set_startpoint");
field53->setAccessType("inputOnly");
field53->setType("SFVec3f");
Script46->addField(field53);

field* field54 = new field();
field54->setName("set_endpoint");
field54->setAccessType("inputOnly");
field54->setType("SFVec3f");
Script46->addField(field54);

IS* IS55 = new IS();
Connect* connect56 = new Connect();
connect56->setNodeField("startnode");
connect56->setProtoField("startnode");
IS55->addConnect(connect56);

Connect* connect57 = new Connect();
connect57->setNodeField("endnode");
connect57->setProtoField("endnode");
IS55->addConnect(connect57);

Connect* connect58 = new Connect();
connect58->setNodeField("set_startpoint");
connect58->setProtoField("set_startpoint");
IS55->addConnect(connect58);

Connect* connect59 = new Connect();
connect59->setNodeField("set_endpoint");
connect59->setProtoField("set_endpoint");
IS55->addConnect(connect59);

Script46->setIS(IS55);


Script46.setSourceCode(`ecmascript:\n"+
"        function recompute(startpoint,endpoint){\n"+
"	    if (typeof endpoint === 'undefined') {\n"+
"		return;\n"+
"	    }\n"+
"            var dif = endpoint.subtract(startpoint);\n"+
"            var dist = dif.length()*0.5;\n"+
"            var dif2 = dif.multiply(0.5);\n"+
"            var norm = dif.normalize();\n"+
"            var transl = startpoint.add(dif2);\n"+
"	    if (typeof Quaternion !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new Quaternion.rotateFromTo(new SFVec3f(0.0,1.0,0.0), norm)\n"+
"		    };\n"+
"	    } else if (typeof SFRotation !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new SFRotation(new SFVec3f(0.0,1.0,0.0),norm)\n"+
"		    };\n"+
"	    } else {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl\n"+
"		    };\n"+
"	    }\n"+
"	}\n"+
"	function recompute_and_route(startpoint, endpoint) {\n"+
"	      var trafo = recompute(startpoint, endpoint);\n"+
"	      if (trafo) {\n"+
"		      position.translation = trafo.translation;\n"+
"		      rotscale.rotation = trafo.rotation;\n"+
"		      rotscale.scale = trafo.scale;\n"+
"	      }\n"+
"	}\n"+
"        function initialize(){\n"+
"            recompute_and_route(startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_startpoint(val,t){\n"+
"            recompute_and_route(val,endnode.translation);\n"+
"        }\n"+
"        function set_endpoint(val,t){\n"+
"            recompute_and_route(startnode.translation,val);\n"+
"        }`)
Group39->addChild(Script46);

ProtoBody38->addChild(Group39);

ProtoDeclare32->setProtoBody(ProtoBody38);

Scene7->addChild(ProtoDeclare32);

ProtoInstance* ProtoInstance60 = new ProtoInstance();
ProtoInstance60->setName("point");
ProtoInstance60->setDEF("G1");
Scene7->addChild(ProtoInstance60);

ProtoInstance* ProtoInstance61 = new ProtoInstance();
ProtoInstance61->setName("point");
ProtoInstance61->setDEF("G2");
Scene7->addChild(ProtoInstance61);

ProtoInstance* ProtoInstance62 = new ProtoInstance();
ProtoInstance62->setName("point");
ProtoInstance62->setDEF("G3");
Scene7->addChild(ProtoInstance62);

ProtoInstance* ProtoInstance63 = new ProtoInstance();
ProtoInstance63->setName("point");
ProtoInstance63->setDEF("G4");
Scene7->addChild(ProtoInstance63);

ProtoInstance* ProtoInstance64 = new ProtoInstance();
ProtoInstance64->setName("x3dconnector");
ProtoInstance64->setDEF("connector1");
fieldValue* fieldValue65 = new fieldValue();
fieldValue65->setName("startnode");
ProtoInstance* ProtoInstance66 = new ProtoInstance();
ProtoInstance66->setUSE("G1");
fieldValue65->addChild(ProtoInstance66);

ProtoInstance64->addFieldValue(fieldValue65);

fieldValue* fieldValue67 = new fieldValue();
fieldValue67->setName("endnode");
ProtoInstance* ProtoInstance68 = new ProtoInstance();
ProtoInstance68->setUSE("G2");
fieldValue67->addChild(ProtoInstance68);

ProtoInstance64->addFieldValue(fieldValue67);

fieldValue* fieldValue69 = new fieldValue();
fieldValue69->setName("set_startpoint");
ProtoInstance64->addFieldValue(fieldValue69);

fieldValue* fieldValue70 = new fieldValue();
fieldValue70->setName("set_endpoint");
ProtoInstance64->addFieldValue(fieldValue70);

Scene7->addChild(ProtoInstance64);

ProtoInstance* ProtoInstance71 = new ProtoInstance();
ProtoInstance71->setName("x3dconnector");
ProtoInstance71->setDEF("connector2");
fieldValue* fieldValue72 = new fieldValue();
fieldValue72->setName("startnode");
ProtoInstance* ProtoInstance73 = new ProtoInstance();
ProtoInstance73->setUSE("G1");
fieldValue72->addChild(ProtoInstance73);

ProtoInstance71->addFieldValue(fieldValue72);

fieldValue* fieldValue74 = new fieldValue();
fieldValue74->setName("endnode");
ProtoInstance* ProtoInstance75 = new ProtoInstance();
ProtoInstance75->setUSE("G3");
fieldValue74->addChild(ProtoInstance75);

ProtoInstance71->addFieldValue(fieldValue74);

fieldValue* fieldValue76 = new fieldValue();
fieldValue76->setName("set_startpoint");
ProtoInstance71->addFieldValue(fieldValue76);

fieldValue* fieldValue77 = new fieldValue();
fieldValue77->setName("set_endpoint");
ProtoInstance71->addFieldValue(fieldValue77);

Scene7->addChild(ProtoInstance71);

ProtoInstance* ProtoInstance78 = new ProtoInstance();
ProtoInstance78->setName("x3dconnector");
ProtoInstance78->setDEF("connector3");
fieldValue* fieldValue79 = new fieldValue();
fieldValue79->setName("startnode");
ProtoInstance* ProtoInstance80 = new ProtoInstance();
ProtoInstance80->setUSE("G1");
fieldValue79->addChild(ProtoInstance80);

ProtoInstance78->addFieldValue(fieldValue79);

fieldValue* fieldValue81 = new fieldValue();
fieldValue81->setName("endnode");
ProtoInstance* ProtoInstance82 = new ProtoInstance();
ProtoInstance82->setUSE("G4");
fieldValue81->addChild(ProtoInstance82);

ProtoInstance78->addFieldValue(fieldValue81);

fieldValue* fieldValue83 = new fieldValue();
fieldValue83->setName("set_startpoint");
ProtoInstance78->addFieldValue(fieldValue83);

fieldValue* fieldValue84 = new fieldValue();
fieldValue84->setName("set_endpoint");
ProtoInstance78->addFieldValue(fieldValue84);

Scene7->addChild(ProtoInstance78);

ROUTE* ROUTE85 = new ROUTE();
ROUTE85->setFromNode("G1");
ROUTE85->setFromField("translation");
ROUTE85->setToNode("connector1");
ROUTE85->setToField("set_startpoint");
Scene7->addChild(ROUTE85);

ROUTE* ROUTE86 = new ROUTE();
ROUTE86->setFromNode("G2");
ROUTE86->setFromField("translation");
ROUTE86->setToNode("connector1");
ROUTE86->setToField("set_endpoint");
Scene7->addChild(ROUTE86);

ROUTE* ROUTE87 = new ROUTE();
ROUTE87->setFromNode("G1");
ROUTE87->setFromField("translation");
ROUTE87->setToNode("connector2");
ROUTE87->setToField("set_startpoint");
Scene7->addChild(ROUTE87);

ROUTE* ROUTE88 = new ROUTE();
ROUTE88->setFromNode("G3");
ROUTE88->setFromField("translation");
ROUTE88->setToNode("connector2");
ROUTE88->setToField("set_endpoint");
Scene7->addChild(ROUTE88);

ROUTE* ROUTE89 = new ROUTE();
ROUTE89->setFromNode("G1");
ROUTE89->setFromField("translation");
ROUTE89->setToNode("connector3");
ROUTE89->setToField("set_startpoint");
Scene7->addChild(ROUTE89);

ROUTE* ROUTE90 = new ROUTE();
ROUTE90->setFromNode("G4");
ROUTE90->setFromField("translation");
ROUTE90->setToNode("connector3");
ROUTE90->setToField("set_endpoint");
Scene7->addChild(ROUTE90);

X3D0->setScene(Scene7);

X3D0->toXMLString();
}
